colcon_core/task/python/build.py

```diff
@@ -186,16 +186,31 @@ def _undo_install(self, pkg, args, setup_py_data, python_lib):
 
     def _symlinks_in_build(self, args, setup_py_data):
         items = ['setup.py']
+        renamed_items = []
         # add setup.cfg if available
         if os.path.exists(os.path.join(args.path, 'setup.cfg')):
             items.append('setup.cfg')
         # add all first level packages
         package_dir = setup_py_data.get('package_dir') or {}
-        for package in setup_py_data.get('packages') or []:
+        packages = setup_py_data.get('packages') or []
+        for package in packages:
             if '.' in package:
                 continue
             if package in package_dir:
                 items.append(package_dir[package])
+                renamed_items.append((package_dir[package], package))
+                if package_dir[package] in package_dir:
+                    package_dir_package = package_dir[package]
+                    raise RuntimeError(
+                        "The package_dir contains a mapping from '{package}' "
+                        "to '{package_dir_package}' which is also a key"
+                        .format_map(locals()))
+                if package_dir[package] in packages:
+                    package_dir_package = package_dir[package]
+                    raise RuntimeError(
```

Comment:
This definitely improves the support for `package_dir`. It doesn't support the following case, but I'm not sure the following case is worth supporting. Raising an error here is definitely better than silently doing the wrong thing like `python setup.py develop` does.

```
alice
├── dee
│   └── __init__.py
├── dum
│   └── __init__.py
└── setup.py
```

```python
from setuptools import setup

setup(
    name='alice',
    version='0.1.0',
    packages=['dee', 'dum'],
    package_dir={'dee': 'dum', 'dum': 'dee'},
    install_requires=['setuptools'],
)
```

With `python setup.py install` the two modules are swapped

```
>>> import dee
originally dum
>>> import dum
originally dee
```

With `python setup.py develop` the modules are incorrectly not swapped, and no error

```
>>> import dee
originally dee
>>> import dum
originally dum
```

with `colcon build --symlink-install` it raises an error about this case not being supported.

```
Traceback (most recent call last):
  File "/tmp/colcon_pr_test/colcon_ws/colcon-from-source/install/colcon-core/lib/python3.6/site-packages/colcon_core/executor/__init__.py", line 91, in __call__
    rc = await self.task(*args, **kwargs)
  File "/tmp/colcon_pr_test/colcon_ws/colcon-from-source/install/colcon-core/lib/python3.6/site-packages/colcon_core/task/__init__.py", line 93, in __call__
    return await task_method(*args, **kwargs)
  File "/tmp/colcon_pr_test/colcon_ws/colcon-from-source/install/colcon-core/lib/python3.6/site-packages/colcon_core/task/python/build.py", line 88, in build
    self._symlinks_in_build(args, setup_py_data)
  File "/tmp/colcon_pr_test/colcon_ws/colcon-from-source/install/colcon-core/lib/python3.6/site-packages/colcon_core/task/python/build.py", line 207, in _symlinks_in_build
    .format_map(locals()))
RuntimeError: The package_dir contains a mapping from 'dee' to 'dum' which is also a key
```

Reply:
> it raises an error about this case not being supported.

Yeah, this is the best we can do (afaik).